### ISSUE TITLE
Workflow examples with Flux?

### ISSUE THREAD

**Issue by vsoch** [LLNL/maestrowf#416]
Hiya! I'm wondering if you know of any workflow examples that use Maestro and Flux? And ideally has a bit of complexity beyond "hello world!" I am testing Maestro with the Flux Operator, and just had my first view of it via Merlin, which adds a redis/rabbitmq service, and I'm curious to try something without Merlin but still using maestro (to learn better how things work). Thanks for the help!

**Comment by vsoch** [comment]
Oh that plot is so cool! And I can zoom and move around - wow! (it looked like YAML in my email then boum! it's gorgeous). :heart_eyes: 

> So, that is a bit of a work in progress currently. A few fixes in this PR should remove the remaining roadblocks to using it in Maestro: https://github.com/LLNL/maestrowf/pull/415. 

I can work from that branch.

> It'll still be a little convoluted depending on the machine/setup if you need to bootstrap it yourself and construct the uri manually (i.e. running under lsf/slurm rather than fully flux controlled machine), 

For my context, I start with a single container and `flux start --test-size=4` or I move into docker-compose if I want to add another node or a service. The ultimate (goal) destination is the Flux Operator (in Kubernetes) which brings up a set of networked pods running flux, and with the broker started (this is called a MiniCluster). I'm not an expert, but I've set up flux enough times that (at least I hope) I won't have any issue with this.

So maybe I can start with hello-world, and then if/when that works, we could figure out something more complex? I'll get started and ping you with updates here, if it's OK we leave the issue open.

**Comment by vsoch** [comment]
okay I think I found the workflow! Here is my yaml:
```yaml
description:
    name: hello_world_multiparam
    description: A simple multiparameter 'Hello World' study.

batch:
    type: flux
    flux_uri: local:///var/run/flux/local

env:
    variables:
        OUTPUT_PATH: ./studies/hello_world
    labels:
        OUT_FORMAT: $(GREETING)_$(NAME).txt

study:
    - name: hello
      description: Say hello to someone!
      run:
          cmd: |
            $(LAUNCHER) echo "$(GREETING), $(NAME)!" > $(OUT_FORMAT)

global.parameters:
    NAME:
        values: [Pam, Jim, Michael, Dwight]
        label: NAME.%%
    GREETING:
        values: [Hello, Ciao, Hey, Hi]
        label: GREETING.%%
```
I have the flux broker running at that path. I can run the workflow:

```bash
$ maestro run studies/hello-world-multiparam-flux.yaml
```
And then it looks OK:
![image](https://user-images.githubusercontent.com/814322/227076244-60af0f03-79ed-4e17-aaec-8ef823a1ce67.png)

But I don't see that any flux jobs were submit:

```bash
root@dba6b7fb3499:/workflow# flux jobs -a
       JOBID USER     NAME       ST NTASKS NNODES     TIME INFO
```
Do I need to add more parameters to the study block for it to work? https://maestrowf.readthedocs.io/en/latest/schedulers.html#scheduling-studies-a-k-a-the-batch-block

**Comment by vsoch** [comment]
I think I got the lulesh example running (at least submitting jobs!) 
![image](https://user-images.githubusercontent.com/814322/227077861-1bae79e4-6449-4b85-8936-64ebe61ea334.png)

I think it killed my computer the first time - was that a one off thing or should I adjust the example to be smaller in some way? I'm guessing it's expected to run on an actual cluster!

**Comment by vsoch** [comment]
okay I made it smaller - it looks like it was running OK

![image](https://user-images.githubusercontent.com/814322/227078857-356e049b-7ba3-4161-9738-dbdea6e9a8a8.png)

But I think it might have been cancelled right at 2 minutes? (they all were cancelled like this):

![image](https://user-images.githubusercontent.com/814322/227078934-f65f9262-d41d-4f66-b3f4-3162517af836.png)

I see there is a walltime in the yaml file - I'll try increasing that to a large value (assuming my computer is just slow compared to a cluster).

**Comment by vsoch** [comment]
Oh I think that worked! 
![image](https://user-images.githubusercontent.com/814322/227080139-783b7271-33e9-4c7f-ac6d-2136cfd9f9a5.png)

Here is what looks like an output file?
```console
# cat studies/lulesh/lulesh_sample1_20230323-015601/run-lulesh/ITER.10.SIZE.100/SIZE.100.ITER.10.log 
Running problem size 100^3 per domain until completion
Num processors: 1
Num threads: 2
Total number of elements: 1000000 

To run other sizes, use -s <integer>.
To run a fixed number of iterations, use -i <integer>.
To run a more or less balanced region set, use -b <integer>.
To change the relative costs of regions, use -c <integer>.
To print out progress, use -p
To write an output file for VisIt, use -v
See help (-h) for more options

cycle = 1, time = 1.910718e-07, dt=1.910718e-07
cycle = 2, time = 4.203581e-07, dt=2.292862e-07
cycle = 3, time = 4.989486e-07, dt=7.859057e-08
cycle = 4, time = 5.645779e-07, dt=6.562923e-08
cycle = 5, time = 6.230838e-07, dt=5.850593e-08
cycle = 6, time = 6.774092e-07, dt=5.432541e-08
cycle = 7, time = 7.290587e-07, dt=5.164946e-08
cycle = 8, time = 7.789464e-07, dt=4.988778e-08
cycle = 9, time = 8.276916e-07, dt=4.874515e-08
cycle = 10, time = 8.757493e-07, dt=4.805775e-08
Run completed:
   Problem size        =  100
   MPI tasks           =  1
   Iteration count     =  10
   Final Origin Energy =  2.596764e+08
   Testing Plane 0 of Energy Array on rank 0:
        MaxAbsDiff   = 8.847564e-09
        TotalAbsDiff = 8.892812e-09
        MaxRelDiff   = 7.277895e-11

Elapsed time         =      2e+02 (s)
Grind time (us/z/c)  =  19.746899 (per dom)  ( 197.46899 overall)
FOM                  =  50.640863 (z/s)
```

**Comment by FrankD412** [comment]
Yep, that's working! Sorry I couldn't chime in earlier, but yeah the wall time is deliberately left small to show the timeout nature of larger problem sizes. This issue makes me want to revisit the docker file packed with Maestro to update it. Looks like you've got it well in hand at this point though. 🙂

**Comment by vsoch** [comment]
Yeah! So maybe another work day (I know it's late) we could brainstorm ideas for how to increase the complexity a bit. I'm thinking something that uses lulesh? E.g., two components, or a component with a service? I don't actually run a lot of workflows (and the ones I used to run were biosciences) so I'm not experienced at what kinds of stuff people actually are interested to run.

Thanks for your help today!

**Comment by FrankD412** [comment]
Sounds good -- I'll give all the credit to @jwhite242 on this one, I didn't do much in the way of help on this one.

And yeah -- just feel free to post when you're ready to start something more complex. 😄

**Comment by vsoch** [comment]
I'm done with the maestro-lulesh example (in docker compose) https://github.com/rse-ops/flux-hpc/tree/main/maestro-lulesh so I'm ready to chat about both of the following, when you have time!

- a reasonable way to do a maestro run + status (and wait for everything to finish before that command exits). Even a custom bash script with some loop fu would work for now.
- a more complex workflow, either using lulesh + something else, a simulation / real world use case, and/or something with a service.

For the first bullet, I could add an interactive example to the operator, but it would be much better to have an example that can launch one script with flux start and have it run to completion. For the second bullet, the context here is we are looking for three solid workflows for a flux operator paper.

**Comment by jwhite242** [comment]
Yeah, the uri can be grabbed that way if you're starting from within the allocation; I've testing with a recipe to submit it to allocations from the login nodes, though that's a little convoluted on the LSF systems with the default flux install since the lsf:jobid resolver isn't in that version :(.

And yes, do be careful with the lulesh sizes: that's size of a 3d volume, so scaling goes up pretty quick!  My laptop similarly didn't like the larger sizes (you can tweak the parameter values in the global.parameters block; there's not really a particular reason for those numbers, so twiddle with them as much as you like).

Also something i've noticed you may need to turn on with some of the parallel problems: using the 'nested: True' option in the parallel step keys, goes where procs/nodes/etc are.  Running the test spec below i noticed it double counting resources without it for some reason:

``` yaml
description:
    name: hello_bye_world_flux
    description: A study that says hello and bye to multiple people.

batch:
    type: flux
    host: rzansel
    bank: guests
    queue: pdebug
    force_broker: True
    # flux_uri: "ssh://rzansel10/var/tmp/flux-sqPdPn/local-0"

env:
    variables:
        OUTPUT_PATH: ./sample_output/hello_world
    labels:
        OUT_FORMAT: $(GREETING)_$(NAME).txt

study:
    - name: hello_world
      description: Say hello to someone!
      run:
          cmd: |
            echo "$(GREETING), $(NAME)!" > $(OUT_FORMAT)
            $(LAUNCHER) sleep 30
          procs: 1
          nodes: 1
          cores per task: 1
          nested: True
          walltime: "00:60"
          
          
    - name: bye_world
      description: Say bye to someone!
      run:
          cmd: |
            echo "Bye, World!" > bye.txt
            $(LAUNCHER) sleep 30
          procs: 1
          nodes: 1
          cores per task: 1
          nested: True
          walltime: "00:60"
          depends: [hello_world]

global.parameters:
    NAME:
        values: [Pam, Jim, Michael, Dwight]
        label: NAME.%%
    GREETING:
        values: [Hello, Ciao, Hey, Hi]
        label: GREETING.%%
```

**Comment by vsoch** [comment]
Do you have any suggestions for a completely headless interaction? E.g., a script to submit, wait for completion, and then maybe I could cat some results file at the end.

**Comment by FrankD412** [comment]
> * a reasonable way to do a maestro run + status (and wait for everything to finish before that command exits). Even a custom bash script with some loop fu would work for now.

For the waiting bit, `maestro run -fg ..` runs Maestro in the foreground which forces the command line to hold for completion. If memory serves, that also outputs the logging that Maestro produces to stdout. That could be used to hold the command, especially if you pipe it to `/dev/null` to hide the output (unless you'd like the printing).

Is that kind of what you're looking for?

**Comment by vsoch** [comment]
I'm running into some trouble on the Flux Operator - both running as root or the flux user (which works in a single container). Why would my flux batch type be using a localscriptexecutor? 

![image](https://user-images.githubusercontent.com/814322/227344158-9acf9115-60cf-4ea3-966b-37d7ef486d93.png)

The permission denied is confusing because the shared storage is owned by the running user - and if I shelled in and manually ran it there isn't an issue. So something is up (that I can't figure out).

**Comment by jwhite242** [comment]
So, the local script adapter is what runs anything that maestro doesn't think is scheduled: so look for steps that don't use $(LAUNCHER) or are missing any of the resource keys like nodes/procs.  You can certainly add those to that make step to force it to be scheduled which may be the better option even if not doing a parallel make.  Copying over from the run step (not sure what the wall time needs to be on that compile step):

``` yaml
    - name: make-lulesh
       description: ...
       run:
          cmd: |
             ...
          nodes: 1
          procs: 1
          cores per task: 1
          nested: True
          walltime: "00:02:00
```

That permissions issue is interesting though.   There anything off between the permissions on that file and the user that's running the conductor process?

**Comment by vsoch** [comment]
I see the problem - so the command that runs subprocess.Popen normally has a command that looks correct (e.g., for git)

```
In [1]: cmd
Out[1]: 
['git',
 'clone',
 'https://github.com/LLNL/LULESH.git',
 '/workflow/shared/studies/lulesh/lulesh_sample1_20230323-203813/LULESH']
```
But for whatever reason, when it hits the same function with the intention to execute that script, there is no executable. It's treating the script as an executable, and not even in a list (which doesn't work either).

```
In [1]: cmd
Out[1]: '/workflow/shared/studies/lulesh/lulesh_sample1_20230323-203813/make-lulesh/make-lulesh.flux.sh'
```
If I give this to popen, I reproduce that permissions error. What I need to do is provide a shell executor proper, e.g.,:

```bash
cmd = ['/bin/bash',  cmd]
```
That works as I'd expect!

![image](https://user-images.githubusercontent.com/814322/227354830-b7baa586-5a4d-43d9-b08d-61e26ed41356.png)

I'm going to go backwards and see why the script is generated like that. Arguably there should be an actual shell executor (bash or sh) added as the first part of the command somewhere?

**Comment by vsoch** [comment]
I think what we need here: https://github.com/LLNL/maestrowf/blob/51056ccfe279495a329036597510d1489dd0c1b1/maestrowf/interfaces/script/localscriptadapter.py#L136 is some ability to explicitly define an executable for a script. I understand why you did it - you want to be able to just run a random path with subprocess (and not know what kind it is) but (for reasons I don't understand) that isn't working in my environment. Could there be a way to define the executable in the step, maybe like:

```yaml
study:
    - name: make-lulesh
      description: Build the MPI enabled version of LULESH.
      executable: /bin/bash
      run:
          cmd: |
            cd $(LULESH) && mkdir build && cd build && cmake -WITH_MPI=Off -WITH_OPENMP=Off .. && make
          depends: []
```
This could be useful for other reasons too - e.g., think of the "too many python installs" problem - maybe I as a user want to target a specific one.

**Comment by jwhite242** [comment]
Ok, so your digging reminded me of something we fixed in the lsf adapter recently that fixes this:

When the adapter determines this step shouldn't be scheduled, instead of writing the scheduler batch file header (all the #SBATCH junk) it is supposed to write the bash shebang there like this:
https://github.com/LLNL/maestrowf/blob/51056ccfe279495a329036597510d1489dd0c1b1/maestrowf/interfaces/script/lsfscriptadapter.py#L471

But the flux adapter looks to have the same bug where it missed that part and just put /bin/bash there (the self._exec is what sets the executable): https://github.com/LLNL/maestrowf/blob/51056ccfe279495a329036597510d1489dd0c1b1/maestrowf/interfaces/script/fluxscriptadapter.py#L316

**Comment by vsoch** [comment]
That might definitely need to be fixed! But it's a different issue.

I think there is something going on my my shared volume - it's giving me permission denied for a lot of executables. E.g.,:

![image](https://user-images.githubusercontent.com/814322/227365005-3c307ade-b183-4cac-b679-08acafb9c7c6.png)

So likely this doesn't have to do with maestro. I just wish I had ideas why it's doing that!

**Comment by vsoch** [comment]
lol! So I mounted the volume in the vm to my host, hoping it would give it some magical inherited filesystem attributes/permissions. And then it all worked. :sob: :laughing: 

![image](https://user-images.githubusercontent.com/814322/227370510-6f62aae5-9989-450a-98ca-9c2cfd4fb692.png)

I hadn't been bitten by this bug yet with a volume in the VM being mounted to pods - I bet it's because the other workflows are just writing / reading data files and not executing anything.  I'm running it fully headless now with the operator - first with root, and then I'm hoping I can fall back to the fluxuser (which worked for the single docker container example). Fingers crossed for everything to work!

I haven't (yet) run into the flux script adapter issue, but if you think it's an issue I can offer to open a PR to fix it - it's the least I can do for taking up your time today on this!

**Comment by vsoch** [comment]
okay - all things a success! https://asciinema.org/a/569620?speed=2 The minicluster.yaml will be merged soon. So final items for this issue:

1. would you like me to open a PR to fix the bug you found (is it a bug?)
2. can we chat about how to make a more complex workflow? Maybe something with a service or just a larger simulation we can easily get performance metrics for?

**Comment by vsoch** [comment]
hey @FrankD412 @jwhite242 ! We chat with some folks from the workflows teams today, and I wanted to follow up here on the questions above. For the first, likely if I didn't run into the bug it doesn't need fixing. For the second, I wanted to see if there were ideas for making a more complex workflow. Specifically, have you heard of workflow tools that use "wrappers" or (in simpler terms) modules that are installable from git or similar, that basically come down to shell / python / similar scripts you can provide parameters to? Here is an example with snakemake https://snakemake-wrappers.readthedocs.io/en/stable/ and their main "wrappers" repository (an example): https://github.com/snakemake/snakemake-wrappers/tree/master/bio/bcftools/call.

I'm wondering if we can define common needs for maestro workflows (or patterns) if we could think about designing something similar, and then assembling workflows like blocks. The core question for this to make sense is if HPC folks have common things they want to run (e.g., imagine if there would be modules for different means to process data, or something like running lammps). My thinking is (longer term) if maestro and merlin become the best practice for the lab, possibly other labs, and a larger open source community, it would be nice to have a similar wrapper / block / module setup for steps in that they could be shared between workflows and tested in isolation.

Let me know your thoughts! I haven't developed for either of these tools (yet) but very likely I could help out.

**Comment by jwhite242** [comment]
So I've got a hunch on why you didn't run into the bug: the process i"m using is trying to hook maestro up to the system flux instances rather than installing flux and maestro with the same python (be it system or virtualenv).  Doing it this way vs say using the docker image (if that's what you tested?), the flux job objects get flagged as un-pickleable as they're actually cffi critters.   Another aspect of this fix is starting to lay the ground work for submitting to flux native machines which means lifting the URI requirement in the flux adapters (usually just the one exported by flux when you start an allocation, but optionally manually in the spec).   Let me know if this is what you were doing, or if there was some other install procedure you used; still working on the flux docs in maestro and would love to include your process there too (and feel free to add recipes into the docs yourself too, contributions are very welcome here!).

Also what might you be looking for in terms of more complex workflows?   There's certainly more interesting topology's that could be constructed with maestro specs than what's in the tutorials now (multiple funnel steps, or expanding back into parameters after funnel steps..).   Having some more things like those wrappers in the maestro sheet music project or something similar could be pretty interesting.  There are some efforts internally that basically make some Maestro spec generators for specific model sets, though even those are fairly project specific being tied to specific simulation codes and more application dependent processing.  Some general curve processing bits are ~common and there's work going on there for some more sharable/reusable pieces that we could chat more about, but beyond that it rapidly devolves into more bespoke collections of things including visit powered data extraction that's highly problem dependent.  

There is an interesting notion of being able to do something related that we've been kicking around: maestro run <git url>.  Though that can run into a bit of a bootstrapping problem given parameter generation packages/etc might need setting up before you could even run such a thing;  similarly, some library of pgen tools could be interesting if we could make the parameter names in them a little easier to swap out on the command line..    

The lulesh example is probably the closest one to something like your running lammps example (having examples for that would be super cool too; not sure if we want to start building stuff like that out in sheetmusic to decouple lammps runnign bugs from the maestro version the sample is obtained from); that's a pretty reasonable example of what simulation steps can look like, with varying complexity based on the code and it's input config or any input config building it might get packaged with.  THink we can probalby come up with some more interesting mfem based examples too.  Would be great to get some more application problem proxies in here for sure.
